### PR TITLE
Automatically generate project specific derived data path

### PR DESCRIPTION
**Pull request by lpusok**
### Checklist

- [x] I've read and followed the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- [x] `step.yml` and `README.md` is updated with the changes (if needed)

### Version

Requires a *MINOR* [version update](https://semver.org/)

### Context

Added *Xcode Project Path* (`project_path`) Input. This allows automatic selection of the correct project-specific DerivedData path. The default behavior does not change, to use this option the *Derived Data Path* (`derived_data_path`) needs to be manually cleaned.

### CODE REVIEW

File: step/step.go [hunk 1]
```diff
@@ -42,40 +51,68 @@ func New(
...
 	stepconf.Print(input)
+	step.logger.EnableDebugLog(input.Verbose)
+
+	input.DerivedDataPath = strings.TrimSpace(input.DerivedDataPath)
+	input.ProjectPath = strings.TrimSpace(input.ProjectPath)
+	if input.DerivedDataPath == "" && input.ProjectPath == "" {
+		return Config{}, fmt.Errorf("provide either Derived Data Path (derived_data_path) or Xcode Project Path (project_path) Inputs")
+	}
+	if input.DerivedDataPath != "" && input.ProjectPath != "" {
+		step.logger.Warnf("Both Derived Data Path (derived_data_path) and Xcode Project Path (project_path) Inputs are provided, only derived_data_path is used, project_path is ignored")
```
Comment (godrei) [inline step/step.go:82]: Do we really ignore `input.ProjectPath` in this case? In the line 85 it seems we use the project path when provided.

File: step/step.go [hunk 2]
```diff
@@ -42,40 +51,68 @@ func New(
...
+	}
 
-	path := filepath.Join(input.DerivedDataPath, "SourcePackages")
+	sourcePackagesPath := filepath.Join(input.DerivedDataPath, "SourcePackages")
+	if input.ProjectPath != "" {
+		var err error
+		if input.ProjectPath, err = step.pathModifier.AbsPath(input.ProjectPath); err != nil {
+			return Config{}, fmt.Errorf("failed to expand absolute project path: %w", err)
```
Comment (godrei) [inline step/step.go]: failed to expand absolute project path -> failed to expand project path

File: step/step.go [hunk 3]
```diff
@@ -42,40 +51,68 @@ func New(
...
+		}
+		// project specific path already contains SourcePacages ($HOME/Library/Developer/Xcode/DerivedData/[PER_PROJECT_DERIVED_DATA]/SourcePackages)
+		if sourcePackagesPath, err = step.derivedDataPathProvider.SwiftPackagesPath(input.ProjectPath); err != nil {
+			return Config{}, fmt.Errorf("failed to get Derived Data Path: %w", err)
+		}
+	}
+
+	return Config{
+		CachePaths: sourcePackagesPath,
+		IsVerbose:  input.Verbose,
```
Comment (godrei) [inline step/step.go:99]: Do we need to pass down the IsVerbose option? Is it used later on?

File: step/step.go (hunk 1 above)
Comment (lpusok) [reply step/step.go:82]: Good catch, fixed.

File: step/step.go (hunk 2 above)
Comment (lpusok) [reply step/step.go]: fixed

File: step/step.go (hunk 3 above)
Comment (lpusok) [reply step/step.go:99]: It is passed to common cache code in line 114.

**Review by godrei — APPROVED** [review]

**Pull request merged**